linode/manager
# feat: [M3-6147] - Add warning for invalid user data

## Description 📝

The User Data of a linode will generally be used by a something like cloud-init to initialize a VM. There is an extremely broad set of valid configuration formats ([described here](https://cloudinit.readthedocs.io/en/latest/explanation/format.html)). Custom formats are also possible if the user image uses something besides cloud-init.

To helpfully guide Metadata users, this feature uses basic string matching to determine if User Data does not match one of the most common formats and displays a warning to the user. The user is able to ignore the warning and proceed anyways if they wish.

~*Note: Since I built this branch off [M3-6143](https://github.com/Hana-Linode/manager/tree/M3-6143-add-user-data-linode-create), this PR currently includes extra commits until that branch is reviewed and merged.*~

## Preview 📷

https://user-images.githubusercontent.com/122488130/222278964-c0c05f1b-63f7-446f-a539-26c29f04c9b9.mov

## How to test 🧪

[For now, you must be using the dev API.]

1. In the Create Linode flow, select an image that is cloud-init compatible. This will enable the Cloud Data dropdown.
2. In the Cloud Data, verify entering a User Data script that doesn't match a common cloud-config format causes the warning to appear.
3. Verify correcting the format causes the warning to disappear.
